### PR TITLE
My solution chapter4

### PR DESCRIPTION
**Pull request by minmax1996**
### Solutions for Chapter 4
Monads not understood(
cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -259,7 +259,7 @@ name.
 
 > QUESTION: Can you understand why the following implementation of the
   Functor instance for Maybe doesn't compile?
-
+Maybe because in `fmap _ x = x` returns Maybe a insteed of Maybe b
```
Comment (chshersh) [inline src/Chapter4.hs:262]: That's absolutely correct :+1: Function requires to return the value of type `Maybe b`, but you have value of type `Maybe a`

File: src/Chapter4.hs [hunk 2]
```diff
@@ -276,6 +276,13 @@ below. 'Secret' is either an unknown trap with something dangerous
 inside or a reward with some treasure. You never know what's inside
 until opened! But the 'Functor' instance allows changing the reward
 inside, so it is quite handy.
+
+fmap (replicate 3) (Reward "42")
+Reward ["42","42","42"]
+fmap (replicate 3) (Reward 42)
+Reward [42,42,42]
+fmap (replicate 3) (Trap 42)
```
Comment (chshersh) [inline src/Chapter4.hs]: If you like, you can even convert these comments into doctest tests :slightly_smiling_face: 

```suggestion
>>> fmap (replicate 3) (Reward "42")
Reward ["42","42","42"]
>>> fmap (replicate 3) (Reward 42)
Reward [42,42,42]
>>> fmap (replicate 3) (Trap 42)
??? what will be here ???
```

File: src/Chapter4.hs [hunk 3]
```diff
@@ -302,10 +310,20 @@ Implement Functor instance for the "List" type defined below. This
...
 purposes, let's practise our skills on implementing standard
 typeclasses for standard data types.
+myl = Cons 2 $ Cons 1 $ Empty
+fmap (replicate 3) myl
+Cons [2,2,2] (Cons [1,1,1] Empty)
 -}
 data List a
     = Empty
     | Cons a (List a)
+    deriving (Show, Eq)
```
Comment (chshersh) [inline src/Chapter4.hs:321]: Nice idea with deriving `Show` and `Eq` to play with instances in GHCi! :+1:

File: src/Chapter4.hs [hunk 4]
```diff
@@ -469,13 +487,17 @@ Applicatives can be found in many applications:
...
 -}
+
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    Reward f <*> d = fmap f d
```
Comment (chshersh) [inline src/Chapter4.hs:500]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 5]
```diff
@@ -489,7 +511,18 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+append :: List a -> List a -> List a
+append Empty d = d
+append (Cons el l) l2 = Cons el (append l l2)
```
Comment (chshersh) [inline src/Chapter4.hs]: Perfect!

You can one more case to make your solution even more efficient:

```haskell
append xs Empty = xs
```

In that case, you won't need to walk through each element of the first list when the second one is empty and return the list instantly 🚤

File: src/Chapter4.hs [hunk 6]
```diff
@@ -489,7 +511,18 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+append :: List a -> List a -> List a
+append Empty d = d
+append (Cons el l) l2 = Cons el (append l l2) 
 
+instance Applicative (List) where
+    pure :: a -> List a
+    pure i = Cons i $ Empty
```
Comment (chshersh) [inline src/Chapter4.hs]: Dollar is redundant here, no need to pay extra money :slightly_smiling_face: 
```suggestion
    pure i = Cons i Empty 
```

File: src/Chapter4.hs [hunk 7]
```diff
@@ -611,6 +645,15 @@ Implement the 'Monad' instance for our lists.
...
 -}
 
+--copypasted :(
+concatList :: List (List a) -> List a
+concatList Empty = Empty
+concatList (Cons x xs) = append x (concatList xs)
+
+instance Monad (List) where
+    (>>=) :: List a -> (a -> List b) -> List b
+    Empty >>= _ = Empty
```
Comment (chshersh) [inline src/Chapter4.hs]: Actually, you don't need to handle this case separately. `fmap` and `concatList` return `Empty` on `Empty`, so this is already converted :lock:

**Review by chshersh — APPROVED** [review]
Congratulations with finishing all four chapters!
Don't be discouraged by not understanding monads, that's totally normal :ok_hand: 
The topic involves rather new entities, and it usually requires more practice to feel more comfortable with them. If you want to have more practice with monads, I can recommend implement the following functions using monads:

1. Convert two maybes to a single maybe of a pair:
    ```haskell
    maybePair :: Maybe a -> Maybe b -> Maybe (a, b)
    ```
2. Find a sum of two maybes:
    ```haskell
    maybeSum :: Maybe Int -> Maybe Int -> Maybe Int
    ```
3. Can you generalize the above function and notice the pattern? Something like:
    ```haskell
    bind2 :: ??? -> Maybe a -> Maybe b -> Maybe c
    ```
4. Try implementing `smartReplicate` for your `List` using `Monad` instance for lists.

After that, you should be more prepared for Task 8 :slightly_smiling_face:

File: src/Chapter4.hs [hunk 8]
```diff
@@ -652,9 +654,89 @@ concatList (Cons x xs) = append x (concatList xs)
...
+>>> maybePair (Just 1) (Just 2)
+Just (1,2)
+>>> maybePair (Just 1) Nothing
+Nothing
+>>> maybePair Nothing Nothing
+Nothing
+-}
+maybePair :: Maybe a -> Maybe b -> Maybe (a, b)
+maybePair (Just x1) (Just x2) = Just (x1, x2)
+maybePair _ _ = Nothing
```
Comment (chshersh) [inline src/Chapter4.hs]: This is the correct solution :+1: But it doesn't use the `Monad` instance for `Maybe`. I was mostly suggesting these tasks to practice with monads...

Try starting with something like this:

```haskell
maybePair ma mb = ma >>= someFunction
```

and fill the body of `someFunction`. At the beginning, you can implement several helper functions inside `where`. But in the end it should be possible to reduce the implementation to a single one-liner

File: src/Chapter4.hs [hunk 9]
```diff
@@ -652,9 +654,89 @@ concatList (Cons x xs) = append x (concatList xs)
...
+replicate (1 :: Int) :: a -> [a]
+concatList :: List (List a) -> List a
+append :: List a -> List a -> List a
+
+I didnt understand which way it should behave
+>>> smartReplicate 3 (Cons 1 $ Cons 2 $ Empty)
+Cons [1,1,1] (Cons [2,2,2] Empty)
+>>> smartReplicate2 3 (Cons 1 $ Cons 2 $ Empty)
+Cons 1 (Cons 1 (Cons 1 (Cons 2 (Cons 2 (Cons 2 Empty))))
+Nevermind though, I implemented both anyway
```
Comment (chshersh) [inline src/Chapter4.hs]: Yeah, both are correct, but to use the `Monad` instance, you need to have the same type for lists of lists. So it should be either `List (List Int)` or `[[Int]]`

The `Monad` instance for ordinary list implemented with the same meaning as for your `List` data type. If you have a list of values, and a function from an element to a list, then you can map all elements and get a list of lists, and then flatten the lists. 

If you look at the implementation of `smartReplicate` from Chapter Two, you can notice some similarities between `concatMap` and `>>=` for lists. And, indeed, `>>=` for lists is `concatMap`

File: src/Chapter4.hs [hunk 10]
```diff
@@ -668,11 +750,12 @@ or anything else! However, this is not a problem in Haskell. You still
...
 
 Can you implement a monad version of AND, polymorphic over any monad?
-
+(>>=) :: m Bool -> (Bool -> m Bool) -> m Bool
+pure :: Bool -> m Bool
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM fb1 fb2 = fb1 >>= (\r -> if r then fb2 else pure False)
```
Comment (chshersh) [inline src/Chapter4.hs:748]: That's an absolutely correct implementation :+1:

**Review by chshersh — APPROVED** [review]
@minmax1996 It's great to see that you continue learning! Keep up the great work :muscle: 

I left my comments on your solutions for additional tasks. And your solution for the first Advanced task is :top: :100:

**Comment by chshersh** [comment]
@minmax1996 Great solutions! I think the PR for the final Chapter is ready for merge :checkered_flag:

**Comment by chshersh** [comment]
Btw, @minmax1996 Thank you so much for taking this course! :tada: I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform), if you haven't submitted it yet!

<hr>

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

**Comment by minmax1996** [comment]
@chshersh Thank you for great course and attention to my PRs. I leaved feedback, and I'll buy you a Coffee! 😊

**Comment by chshersh** [comment]
@minmax1996 Thank you, your support is much appreciated :hugs:

**Pull request merged**